codegenerator/cli/npm/envio/src/Batch.res

```diff
@@ -2,146 +2,158 @@ type progressedChain = {
   ->ChainMap.values
   ->Js.Array2.some(fetchState => {
-    fetchState->FetchState.isActivelyIndexing &&
-      switch fetchState->FetchState.getEarliestEvent {
-      | Item(_) => true
-      | NoItem(_) => false
-      }
+    fetchState->FetchState.isActivelyIndexing && fetchState->FetchState.hasReadyItem
   })
 }
 
-let popOrderedBatchItems = (
-  ~maxBatchSize,
+let hasMultichainReadyItem = (
+  fetchStates: ChainMap.t<FetchState.t>,
+  ~multichain: InternalConfig.multichain,
+) => {
+  switch multichain {
+  | Ordered => hasOrderedReadyItem(fetchStates)
+  | Unordered => hasUnorderedReadyItem(fetchStates)
+  }
+}
+
+let prepareOrderedBatch = (
+  ~batchSizeTarget,
   ~fetchStates: ChainMap.t<FetchState.t>,
-  ~sizePerChain: dict<int>,
+  ~mutBatchSizePerChain: dict<int>,
 ) => {
+  let batchSize = ref(0)
+  let isFinished = ref(false)
   let items = []
 
-  let rec loop = () =>
-    if items->Array.length < maxBatchSize {
-      switch fetchStates->getOrderedNextItem {
-      | Some({earliestEvent, chain}) =>
-        switch earliestEvent {
-        | NoItem(_) => ()
-        | Item({item, popItemOffQueue}) => {
-            popItemOffQueue()
-            items->Js.Array2.push(item)->ignore
-            sizePerChain->Utils.Dict.incrementByInt(chain->ChainMap.Chain.toChainId)
-            loop()
+  while batchSize.contents < batchSizeTarget && !isFinished.contents {
+    switch fetchStates->getOrderedNextChain(~batchSizePerChain=mutBatchSizePerChain) {
+    | Some(fetchState) => {
+        let itemsCountBefore = switch mutBatchSizePerChain->Utils.Dict.dangerouslyGetByIntNonOption(
+          fetchState.chainId,
+        ) {
+        | Some(batchSize) => batchSize
+        | None => 0
+        }
+        let newItemsCount =
+          fetchState->FetchState.getReadyItemsCount(
+            ~targetSize=batchSizeTarget - batchSize.contents,
+            ~fromItem=itemsCountBefore,
+          )
+
+        if newItemsCount > 0 {
+          for idx in itemsCountBefore to itemsCountBefore + newItemsCount - 1 {
+            items->Js.Array2.push(fetchState.buffer->Belt.Array.getUnsafe(idx))->ignore
           }
+          batchSize := batchSize.contents + newItemsCount
+          mutBatchSizePerChain->Utils.Dict.setByInt(
+            fetchState.chainId,
+            itemsCountBefore + newItemsCount,
+          )
+        } else {
+          isFinished := true
         }
-      | _ => ()
       }
+
+    | None => isFinished := true
     }
-  loop()
+  }
 
   items
 }
 
-let popUnorderedBatchItems = (
-  ~maxBatchSize,
+let prepareUnorderedBatch = (
+  ~batchSizeTarget,
   ~fetchStates: ChainMap.t<FetchState.t>,
-  ~sizePerChain: dict<int>,
+  ~mutBatchSizePerChain: dict<int>,
 ) => {
-  let items = []
-
   let preparedFetchStates =
     fetchStates
     ->ChainMap.values
-    ->FetchState.filterAndSortForUnorderedBatch(~maxBatchSize)
+    ->FetchState.filterAndSortForUnorderedBatch(~batchSizeTarget)
 
-  let idx = ref(0)
+  let chainIdx = ref(0)
```

Comment:
Seem like this means there's stickly load balancing on a chain. Like the first chain in the chain map will always be prioritized? Not sure if there's any problem with this but checking if this is intended behaviour

Reply:
There's `filterAndSortForUnorderedBatch` above to manage load ballancing